Objects/typeobject.c

```diff
@@ -2241,6 +2241,12 @@ extra_ivars(PyTypeObject *type, PyTypeObject *base)
         type->tp_weaklistoffset + sizeof(PyObject *) == t_size &&
         type->tp_flags & Py_TPFLAGS_HEAPTYPE)
         t_size -= sizeof(PyObject *);
+    if (!(type->tp_flags & Py_TPFLAGS_MANAGED_DICT) &&
+        type->tp_dictoffset && base->tp_dictoffset == 0 &&
+        type->tp_dictoffset + sizeof(PyObject *) == t_size &&
+        type->tp_flags & Py_TPFLAGS_HEAPTYPE) {
+        t_size -= sizeof(PyObject *);
+    }
```

Comment:
This reintroduces the old code pre-3.11, but guards it with a check for not `Py_TPFLAGS_MANAGED_DICT`.

https://github.com/python/cpython/commit/8319114feedd2a5b77378bba24eb9fb2689c5033#diff-1decebeef15f4e0b0ce106c665751ec55068d4d1d1825847925ad4f528b5b872L2254

Reply:
The condition is a bit arcane. Could you capture it in a variable or use a macro or inline function so we can name what we are checking for? Currently is a bit difficult to decipher what happens when all of these are true or false

Reply:
So funny thing is I also found the old code arcane and couldn't really understand it well. From my rough guesstimation
```
    // The __dict__ isn't automatically managed by CPython.
    if (!(type->tp_flags & Py_TPFLAGS_MANAGED_DICT) &&
        // That a dictoffset is set. Meaning it's a manual __dict__.
        // And the base class doesn't have a __dict__
        type->tp_dictoffset && base->tp_dictoffset == 0 &&
        // Huh?
        type->tp_dictoffset + sizeof(PyObject *) == t_size &&
        // Not a static type.
        type->tp_flags & Py_TPFLAGS_HEAPTYPE) {
        t_size -= sizeof(PyObject *);
    }
```

Reply:
I think this is a really cursed example of Hyrum's law.

The `extra_ivars` code basically assumes that `__weakreflist__` comes first then `__dict__`. In `mypyc` it has to follow this order to work pre-3.11. Now in 3.11 when we remove that assumption, all the C extensions using this with complicated MROs break 🤦 .